docs/proposals/llm-hyperparameter-optimization-api.md

```diff
@@ -0,0 +1,173 @@
+        - objective_type: Optimization direction for the objective metric, "minimize" or "maximize".
+		- objective_goal: Desired value of the objective metric.
+        - max_trial_count: Maximum number of trials to run.
+        - parallel_trial_count: Number of trials to run in parallel.
+		- max_failed_trial_count: Maximum number of allowed failed trials.
+        - resources_per_trial: Resources required per trial.
+		- pytorch_config: Configuration for PyTorch jobs, including number of workers and processes per worker.
+        - retain_trials: Whether to retain trial resources after completion.
+		- env_per_trial: Environment variables for worker containers.
+        - packages_to_install: Additional Python packages to install.
+        - pip_index_url: URL of the PyPI index for installing packages.
+        """
+        pass  # Implementation logic for initiating the experiment
+```
+
+### Example
+
+```python
+import transformers
+from peft import LoraConfig
+
+from kubeflow.storage_initializer.hugging_face import (
+	HuggingFaceModelParams,
+	HuggingFaceDatasetParams,
+	HuggingFaceTrainerParams,
+)
+
+import kubeflow.katib as katib
+from kubeflow.katib import KatibClient
+
+# Create a Katib client
+katib_client = KatibClient(namespace='kubeflow')
+
+# Run the tuning Experiment
+exp_name = "llm-hp-optimization-test"
+katib_client.tune(
+	name = exp_name,
+	# BERT model URI and type of Transformer to train it.
+	model_provider_parameters = HuggingFaceModelParams(
+		model_uri = "hf://google-bert/bert-base-cased",
+		transformer_type = transformers.AutoModelForSequenceClassification,
+	),
+	# Use 3000 samples from Yelp dataset.
+	dataset_provider_parameters = HuggingFaceDatasetParams(
+		repo_id = "yelp_review_full",
+		split = "train[:3000]",
+	),
+	# Specify HuggingFace Trainer parameters.
+	trainer_parameters = HuggingFaceTrainerParams(
+		training_parameters = transformers.TrainingArguments(
+			output_dir = "test_trainer",
+			save_strategy = "no",
+			eval_strategy = "epoch",
+			disable_tqdm = True,
+			log_level = "info",
+			learning_rate = katib.search.double(min=1e-05, max=5e-05),
+			per_device_train_batch_size = katib.search.int(min=8, max=64),
+			num_train_epochs = katib.search.int(min=1, max=10),
+			weight_decay = katib.search.double(min=0.0, max=1.0),
+		),
+		# Set LoRA config to reduce number of trainable model parameters.
+		lora_config = LoraConfig(
+			r = katib.search.int(min=8, max=32),
+			lora_alpha = 8,
+			lora_dropout = 0.1,
+			bias = "none",
+		),
+	),	
+	objective_metric_name = "eval_accuracy", 
+	objective_type = "maximize", 
+	algorithm_name = "random",
+	max_trial_count = 50,
+	parallel_trial_count = 2,
+	resources_per_trial = {
+		"gpu": 8,
+		"cpu": 20,
+		"memory": "40G",
+	},
+	num_workers = 4,
+	num_procs_per_worker = 2,
+)
+
+# Get the best hyperparameters
+best_hps = katib_client.get_optimal_hyperparameters(exp_name)
+```
+
+## Implementation
```

Comment:
It would be nice to add info how you are going to iterate over `HuggingFaceTrainerParams`, search for HP that user wants to optimize, and do the appropriate substitution in the Trail template.

Reply:
I have added explanation for this part. Please let me know if there are any problems.